feat: refactor `count_pull_requests` to use GraphQL

chazkiker2 opened the pull request:
# Description

This PR improves the speed of the `src/metrics/list_repos` producer by an incredible amount by simple shifting our `count_pull_requests` function to utilize a GraphQL call.

- [x] refactor 'count_pull_requests' function to utilize graphql query

nikomatsakis commented on src/metrics/list_repos.rs:
```diff
@@ -125,3 +126,41 @@ async fn count_pull_requests(org_name: &String, repo_name: &str) -> usize {
+    let thirty_days_ago_str = chars.as_str();
+
+    let query_string = format!(
+        "query {{ search(query:\"repo:{org_name}/{repo_name} is:pr created:>{thirty_days_ago}\", type:ISSUE, last:100) {{ issueCount }} }}",
```
```suggestion
        r#"query {{
            search(query:"repo:{org_name}/{repo_name} is:pr created:>{thirty_days_ago}", type:ISSUE, last:100) {{
                issueCount
            }}
        }}"#,
```

nikomatsakis replied:
might as well add some newlines too

nikomatsakis reviewed (changes requested):
Using raw strings means we don't have to escape the `"`

nikomatsakis commented on src/metrics/list_repos.rs:
```diff
@@ -125,3 +126,41 @@ async fn count_pull_requests(org_name: &String, repo_name: &str) -> usize {
+    let mut chars = s.chars();
+    chars.next_back();
+    chars.next_back();
+    chars.next_back();
```
what on earth is this btw :) I guess shaving off some characters?

nikomatsakis replied:
there has to be a better way...but at least it needs a comment :)

chazkiker2 replied:
haha SO fair.  I'm just trying to get the Date string in the right format. 

It formats like `"2021-05-18UTC"` but we need it like `"2021-05-18"` so I'm removing three chars from the end. 


Very open to suggestions for improvement on how to do this 😆

XAMPPRocky replied:
FWIW since you’re already using chrono you can have it parse it from the string.

```rust
chrono::NaiveDate::parse_from_str(input, "%FUTC") // %F == %Y-%m-%d

chazkiker2 replied:
That's perfect @XAMPPRocky thank you!

chazkiker2 commented:
## Featured Changes
- [x] changed to raw string, as you suggested
- [x] added comments for the anti-idiomatic string manipulation
- [x] added a refactor for `all_repos_graphql`... let me know if you want me to delete that and keep the other `all_repos`

chazkiker2 commented:
maybe don't merge yet — I'd like to clean some things up tomorrow. 

but would love to know whether you like the `all_repos_graphql` or if it adds unnecessary bulk w/ the required `struct`s?

nikomatsakis commented on src/metrics/list_repos.rs:
```diff
@@ -62,66 +64,56 @@ impl Producer for ListReposForOrg {
+            search(
+                query:"repo:{org_name}/{repo_name} is:pr created:>{thirty_days_ago}",
+                type:ISSUE,
+                last:100,
```
I'm confused by this code -- what does `last:100` do? I would expect that to only count up to 100 issues or something

chazkiker2 replied:
This is not an issue for this query in `count_pull_requests_graphql` because we don't actually care about the pages of data (for now). (one of the reasons it's so fast, probably)

Take this query, for example: 
```graphql
{
  search(
    query: "repo:rust-lang/rust is:pr created:>2021-05-19", type: ISSUE, last: 100
  ) {
    issueCount
    pageInfo {
      hasNextPage
      hasPreviousPage
      startCursor
      endCursor
    }
  }
}
```

it returns the following data: 
```json
{
  "data": {
    "search": {
      "issueCount": 588,
      "pageInfo": {
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": "Y3Vyc29yOjE=",
        "endCursor": "Y3Vyc29yOjEwMA=="
      }
    }
  }
}
```

The reason this is such an improvement is that `issueCount` is a native property on whatever GQL type the `search` query responds with. And since that's (for now) the only item we care about—we could go to each next page, but that `issueCount` would stay the same at 588 for each page. Performance is so much better since we're not walking in to the `edges { node {}}` for every single PR.

chazkiker2 replied:
That said, I'll go ahead and change it to `last: 1` in case that helps performance in any marginal way. 

And I'll leave a comment to make the purpose behind this more clear.

nikomatsakis commented on src/metrics.rs:
```diff
@@ -32,3 +33,64 @@ pub use print::Print;
+    let query_string = format!(
+        r#"query {{
+            organization(login:"{org_name}"){{
+                repositories(first:100){{
```
does this assume no more than 100 repos?

chazkiker2 replied:
Huge bug, glad you caught it. 🐛

chazkiker2 commented on src/main.rs:
```diff
@@ -75,7 +77,18 @@ async fn main() {
+            let num_days: i64 = if let Some(number_of_days) = config.github.number_of_days {
+                number_of_days.try_into().unwrap()
+            } else {
+                30
```
Oh cool! Should we make this a Clap arg? I included it in the `Config` instead of `Opt`, but I could see that maybe making more sense as a CLI arg rather than a config property. 

But agreed that this can take place in a separate PR (maybe the one that refactors the command-line args #26?)

nikomatsakis replied:
Yep. I think I'd like to merge the idea of "toml config" and command line options, but I'm not sure the best way to do that. I imagine being able to load options from either place, with the command line taking precedence.

chazkiker2 commented:
### Featured Changes

- cleaned up `date_str` manipulation with @XAMPPRocky 's wonderful suggestion
- fixed pagination for `all_repos_graphql`
- add repo count to terminal output for `list` (I liked it!)
